README.md

```diff
@@ -125,3 +123,16 @@ completes (or fails).
 When the `PipelineRun` has completed, the `taskRuns` field will contain
 references to all `TaskRuns` which were executed and their next and
 previous `TaskRuns`.
+
+### Resources
+
+`Resources` in a pipelines are the set of objects that are going to be used 
+as inputs and outputs of a `TaskRun`. 
+
+* `Resources` is created directly in a pipeline configuration and bound 
+to `TaskRun` as an input and/or output source. 
+* The (optional) `passedConstraint` key on an `input source` defines a set of previous task names.
+* When the `passedConstraint` key is specified on an input source, only the version of 
```

Comment:
Having a hard time wrapping my head around “passedConstraints” bound to “source”
Example: https://github.com/knative/build-pipeline/blob/cf8546cc88b6ee298f99b6f64b63d5db8acb78c2/examples/pipelines/guestbook.yaml#L79
Specifically in this case, we have 1 source key but 2 `passedConstraints`. 
Which source version do i pick?


I see a CI Pipeline defined like this:
Where, sources are input to a pipeline and then something comes out of the pipeline. It could be a production ready image or an app.
Pipeline consists of tasks and each task executes some steps on the source and then produce resources like (images, jars, test results) etc.
The pipeline will be run, managed by a controller.  Let us say we call it PipelineController.

[In my head Pipeline looks like this](https://docs.google.com/document/d/1nIxVqZbOLIPVRGuVvCEw3zn30l5fbvPgbwGrsLJS-dg/edit#heading=h.vgrkr0rpzh9d)
Each PipelineRun will have a RunContext which will store the state of the pipeline.
The PipelineController will add the inputs and outputs produced along the way are added to RunContext along with their versions.
Now, when Task1 completes, the PipelineController should invoke Task2. It will refer the versions of the inputs and outputs in the context and pass the required artifacts.
This behaviour should be “implicit”. 
We do not want different versions of the same source to be running in the same pipeline run.
Adding a `passingConstraint` which is optional at the source level may actually lead to such scenarios. 
Hence, i feel moving `passedConstraint` to task level makes more sense.

Reply:
Thanks for clarifying @pivotal-nader-ziada 

> We are trying to make it explicit that the version of the resource stays the same across multiple tasks in a pipelineRun. By adding the passedConstraint key on the InputSource that means that the input of that task has to be the same version that passed the previous task(s) mentioned in constraint.
This allows for the option of another resource (for example a test only repo) that always wants to get latest in each run and doesn't need the same version

Ok, this makes things more clear now. I think,  we just have defaults flipped.  If you have a resource defined like this, i was more inclined towards overriding the resource version.  
```
resource:
    - name: guestbook
      type: git
      url: github.com/kubernetes/examples
      serviceAccount: githubServiceAccount
      revision: HEAD
```
e.g. The first task in the pipeline pulling in guestbook will use revision `HEAD` and lock its sha. The next tasks in pipeline which depend on resource `guestbook` will override the revision with the locked sha built by 1st task in the pipeline.
I like your approach where you explicitly specify the `passedConstraint`. 
However, we are now relied on pipeline owners to make sure they correctly configure it. We can work around it by adding some "intelligence" in validators and spinning out warning if we see patterns where pipeline-creators are pulling in the same resource in two tasks but not using `passedConstraint`.